multishot/dynamic_state.ml

```diff
@@ -0,0 +1,265 @@
+(* version 4. Uses dynamic creation of new effect names to simulate
+   the creation of new reference cells. Initially, there is only one
+   effect 'New', which can be used to dynamically create new effect
+   names. The handler for 'New' wraps the continuation in a new
+   handler that handles the freshly generated effect names. This setup
+   yields the same interface as ML refs, except that there is no way
+   to compare references for equality. This is because cells are
+   represeted as objects with a pair of a 'write' method and a 'read'
+   method, so it is possible to create new references that reference
+   the same underlying data without the access objects being
+   equal. This is similar to the situation in Idealised Algol, where
+   variables are ways to affect the state, but have no independent
+   existence of their own.
+
+   Compared to the example in "ref.ml", this implementation does not
+   require a universal type, nor does it have "impossible" cases.
+
+   This example also includes an unneccessary extra 'Choice' effect to
+   demonstrate the combination of other effects with state in the same
+   handler. This uses the experimental Obj.clone_continuation function to clone
+   continuations. *)
+type 'a reff =
+  < get : 'a
+  ; put : 'a -> unit
+  ; internals : (module GetPutOps with type t = 'a) >
+
+type _ Effect.t += New : 'a -> 'a reff t
+type _ Effect.t += Choice : bool t
+
+let run4 main =
+  let donew : type a b. (a reff, b) continuation -> a -> b =
+   fun k ->
+    let module Ops =
+      MakeGetPut
+        (struct
+          type t = a
+        end)
+        ()
+    in
+    let cell =
+      object
+        method get = perform Ops.Get
+        method put x = perform (Ops.Put x)
+        method internals = (module Ops : GetPutOps with type t = a)
+      end
+    in
+    match_with (continue k) cell
+      {
+        retc = (fun s _ -> s);
+        exnc = (fun e -> raise e);
+        effc =
+          (fun (type c) (e : c t) ->
+            match e with
+            | Ops.Put v ->
+                Some (fun (k : (c, _) continuation) _ -> continue k () v)
+            | Ops.Get ->
+                Some (fun (k : (c, _) continuation) (s : a) -> continue k s s)
+            | _ -> None);
+      }
+  in
+  try_with main ()
+    {
+      effc =
+        (fun (type a) (e : a t) ->
+          match e with
+          | New v -> Some (fun (k : (a, _) continuation) -> donew k v)
+          | Choice ->
+              Some
+                (fun (k : (a, _) continuation) ->
+                  let k' = Multicont.Deep.clone_continuation k in
+                  continue k true;
+                  continue k' false)
+          | _ -> None);
+    }
+
+let newref i = perform (New i)
+let ( := ) r x = r#put x
+let ( ! ) r = r#get
+
+let test4 () =
+  let a = newref 0 in
+  let b = newref "str" in
+  if perform Choice then (
+    a := String.length !b;
+    b := string_of_int !a;
+    print_endline !b)
+  else print_endline !b
```

Comment:
I suppose for consistency it would be good to include the runner here as well, e.g. `main4 () = run4 test4`.